library/kani_macros/src/derive.rs

```diff
@@ -0,0 +1,144 @@
+        if let GenericParam::Type(type_param) = param {
+            type_param.bounds.push(parse_quote!(kani::Arbitrary));
+        }
+    });
+    generics
+}
+
+/// Generate the body of the function `any()`.
+/// This will create the non-deterministic object.
+/// E.g.:
+/// ```
+/// #[derive(Arbitrary)]
+/// struct Point { x: u8, y: u8 }
+/// ```
+/// will generate the following body for `fn any()`:
+/// ```
+/// fn any() -> Self {
+///    Self { x: kani::any(), y: kani::any() }
+/// }
+/// ```
+fn fn_any_body(ident: &Ident, data: &Data) -> TokenStream {
+    match data {
+        Data::Struct(struct_data) => init_item(ident, &struct_data.fields),
+        Data::Enum(enum_data) => fn_any_enum(ident, enum_data),
+        Data::Union(_) => {
+            abort!(Span::call_site(), "Cannot derive `Arbitrary` for `{}` union", ident;
+                note = ident.span() =>
+                "The derive `Arbitrary` cannot be used for unions such as `{}`", ident
+            )
+        }
+    }
+}
+
+/// Generate an item initialization where an item can be a struct or a variant.
+/// For named fields, this will generate: `Item { field1: kani::any(), field2: kani::any(), .. }`
+/// For unnamed fields, this will generate: `Item (kani::any(), kani::any(), ..)`
+/// For unit field, generate an empty initialization.
+fn init_item(ident: &Ident, fields: &Fields) -> TokenStream {
+    match fields {
+        Fields::Named(ref fields) => {
+            // Use the span of each `syn::Field`. This way if one of the field types does not
+            // implement `Arbitrary` then the compiler's error message underlines which field it
+            // is. An example is shown in the readme of the parent directory.
+            let init = fields.named.iter().map(|field| {
+                let name = &field.ident;
+                quote_spanned! {field.span()=>
+                    #name: kani::any()
+                }
+            });
+            quote! {
+                #ident {#( #init,)*}
+            }
+        }
+        Fields::Unnamed(ref fields) => {
+            // Expands to an expression like
+            // Self(kani::any(), kani::any(), ..., kani::any());
+            let init = fields.unnamed.iter().map(|field| {
+                quote_spanned! {field.span()=>
+                    kani::any()
+                }
+            });
+            quote! {
+                #ident(#( #init,)*)
+            }
+        }
+        Fields::Unit => {
+            quote! {
+                #ident
+            }
+        }
+    }
+}
+
+/// Generate the body of the function `any()` for structures.
+fn fn_any_enum(ident: &Ident, data: &DataEnum) -> TokenStream {
+    if data.variants.is_empty() {
+        abort!(Span::call_site(), "Cannot derive `Arbitrary` for `{}`", ident;
+            note = ident.span() => "`{}` has zero variants and cannot be instantiated", ident
+        );
+    }
+    let arms = data.variants.iter().enumerate().map(|(idx, variant)| {
```

Comment:
I think I see what this is doing, but can you add to the function documentation an example of the output?

Reply:
Done